design/design.md

```diff
@@ -0,0 +1,414 @@
+It is possible that Peras and Leios requirements on the proof size cannot
+be satisfied using ALBA alone.
+In that case, further techniques can be considered such as SNARKs and
+taking the stake distribution into account (todo: see fait-accompli).
+
+These protocols' requirements about communication and computational complexity
+are not currently known.
+However, the communication complexity of the implemented ALBA schemes will not greatly
+exceed that of the simple lottery, and the prover running time is expected to be
+in the order of a second.
+
+### Interfaces
+
+Peras and Leios are implemented in Haskell and thus require bindings to our
+Rust implementation.
+Mithril is implemented in Rust and thus can use our implementation natively.
+
+Beyond this, there are no other clearly stated requirements.
+(todo: section about the solution) describes some design choices for
+the library's interfaces.
+
+## Proposed design choices
+
+This section describes various design choices for our implementation.
+
+### ALBA algorithms
+
+ALBA comes in multiple different flavors.
+Two different setting dimensions, unweighted / weighted and
+centralized / decentralized, define four different ALBA variants.
+See (todo: overview section) for an overview.
+
+There also exist multiple ALBA schemes;
+currently four are known for the centralized unweighted setting:
+1. (basic) Telescope (section 3.1 in the paper),
+2. Telescope with Prehashing (section 3.2 in the paper),
+3. Telescope with prehashing, repetitions and bounded DFS
+("Telescope Bounded" in short, section 3.2.2 in the paper),
+4. simple lottery (section 4.1 in the paper).
+
+The same four schemes can be made decentralized by adding a simple lottery
+(or just changing interfaces in case of the simple lottery which itself is
+already a decentralized scheme), see section 4.1 and 4.2 in the paper.
+Additionally, the support for weights can be introduced in the decentralized
+schemes by having the simple lottery take into account weights, as done in
+Algorand's sortition (see section 5 in the paper).
+(Thus, the simple lottery in the decentralized weighted setting is equivalent
+to Algorand's sortition.)
+Finally, we can create centralized weighted ALBA schemes by remodeling the
+decentralized weighted schemes as centralized;
+this is currently the only known efficient way to implement weights.
+
+We will implement Telescope Bounded,
+as well as, the simple lottery in all four settings (centralized+decentralized,
+unweighted+weighted).
+The following is the full list of eight implemented schemes.
+* Centralized unweighted Telescope Bounded (section 3.2.2)
+* Decentralized unweighted Telescope Bounded (section 3.2.2 + 4.2)
+* Centralized+decentralized weighted Telescope Bounded (section 3.2.2 + 4.2 + 5)
+* Centralized+decentralized unweighted simple lottery (section 4.1)
+* Centralized+decentralized weighted simple lottery (section 4.1 + 5)
+
+#### Discussion
```

Comment:
You could be more concise in the whole section. Many paragraphs can be rewritten with fewer lines and words. You do not needto explain why something is more efficient when it is obvious for instance.

I also find the style a bit incoherent since you first list the reasons "in english" (the following reasons. ...), then you have long sentence to list all for lottery, and finally you have an itemize for the redundancy.